File: Vector/OOPs/OOPsConceptAssignment/Abstraction/AbstractClass.js
// JavaScript does not have built-in abstract classes, but you can create a base class with methods that throw errors if they are not implemented in derived classes.

class Appliance {
    turnOn() {
        throw new Error("This method should be overridden");
    }
}

class WashingMachine extends Appliance {
    turnOn() {
        console.log("Washing machine is now on");
    }
}

class Refrigerator extends Appliance {
    turnOn() {
        console.log("Refrigerator is now on");
    }
}

function main() {
    const washer = new WashingMachine();
    const fridge = new Refrigerator();

    washer.turnOn();
    fridge.turnOn();
}

main();
// Washing machine turned on.
// Refrigerator turned on.

File: Vector/OOPs/OOPsConceptAssignment/Abstraction/AbstractClass.java
// Abstract Class
// Abstraction is a fundamental concept in object-oriented programming (OOP) that 
// focuses on hiding the complex implementation details of a system and 
// exposing only the essential features or functionalities to the users. 
// It allows programmers to work with high-level concepts without worrying about the low-level implementation details.

// Create an abstract class Appliance with an abstract method turnOn().
//  Extend this class with WashingMachine and Refrigerator, 
// each providing an implementation of turnOn(). 
// Write a main method to demonstrate the usage of these classes.


public abstract class Appliance {
    public abstract void turnOn();
}

class WashingMachine extends Appliance {
    @Override
    public void turnOn() {
        System.out.println("Washing machine turned on.");
    }
}

class Refrigerator extends Appliance {
    @Override
    public void turnOn() {
        System.out.println("Refrigerator turned on.");
    }
}

class Main {
    public static void main(String[] args) {
        Appliance washingMachine = new WashingMachine();
        washingMachine.turnOn();

        Appliance refrigerator = new Refrigerator();
        refrigerator.turnOn();
    }
}

// Washing machine turned on.
// Refrigerator turned on.
